File: py/src/configu/model/config_schema.py
import functools
import json
import re
from dataclasses import dataclass
from itertools import cycle
from json import JSONDecodeError
from pathlib import Path
from typing import Dict, Callable, List

import pyvalidator

from .generated import ConfigSchema as IConfigSchema, Cfgu, ConfigSchemaType, CfguType, from_dict
from ..utils import error_message, is_valid_name


@dataclass
class ConfigSchemaDefinition:
    NAME: str = 'cfgu'
    EXT: str = '.cfgu'
    VALIDATORS: Dict[str, Callable[[str], bool]] = None

    def __post_init__(self):
        if self.VALIDATORS is None:
            self.VALIDATORS = {
                "Boolean": pyvalidator.is_boolean,
                "Number": pyvalidator.is_number,
                "String": lambda value: isinstance(value, str),
                "UUID": pyvalidator.is_uuid,
                "SemVer": pyvalidator.is_semantic_version,
                "Email": pyvalidator.is_email,
                "MobilePhone": pyvalidator.is_mobile_number,
                "LatLong": pyvalidator.is_lat_long,
                "Color": lambda value: (
                    pyvalidator.is_hexadecimal(value)
                    or pyvalidator.is_hsl(value)
                    or pyvalidator.is_rgb_color(value)
                ),
                "IPv4": lambda value: pyvalidator.is_ip(value, 4),
                "IPv6": lambda value: pyvalidator.is_ip(value, 6),
                "Domain": pyvalidator.is_fqdn,
                "URL": pyvalidator.is_url,
                "ConnectionString": lambda value: re.fullmatch(
                    r"^(?:([^:/?#\s]+):/{2})?(?:([^@/?#\s]+)@)?([^/?#\s]+)?(?:/([^?#\s]*))?(?:[?]([^#\s]+))?\S*$",
                    value, re.RegexFlag.M) is not None,
                "Hex": pyvalidator.is_hexadecimal,
                "Base64": pyvalidator.is_base64,
                "MD5": pyvalidator.is_md5,
                "SHA": lambda value: (
                    pyvalidator.is_hash(value, 'sha1')
                    or pyvalidator.is_hash(value, 'sha256')
                    or pyvalidator.is_hash(value, 'sha384')
                    or pyvalidator.is_hash(value, 'sha512')
                ),
                "Currency": pyvalidator.is_currency,
            }

    @functools.cached_property
    def ext(self) -> str:
        return " | ".join(["".join(ext) for ext in zip(cycle(self.EXT), self.types.keys())])

    @functools.cached_property
    def types(self) -> Dict[str, ConfigSchemaType]:
        return {f'.{schema_type.value}': schema_type for schema_type in ConfigSchemaType}

    @functools.cached_property
    def props(self) -> List[str]:
        return list(Cfgu.__annotations__.keys())


class ConfigSchema(IConfigSchema):
    """"""
    # todo nothing is done with PROPS.. ?
    #  why is this here anyway? i guess there will be other Schema types?
    #  if so this needs elevation or better if ConfigSchemaType will contain all this. if not its redundant.
    _SchemaDefinition = ConfigSchemaDefinition()

    def __init__(self, path: str) -> None:
        error_location = [self.__class__.__name__, self.__init__.__name__]
        if re.match(rf'.*({ConfigSchema._SchemaDefinition.ext})', path) is None:
            raise ValueError(error_message(f"invalid path {path}", error_location,
                                           f"file extension must be {ConfigSchema._SchemaDefinition.ext}"))
        super().__init__(path=path, type=ConfigSchema._SchemaDefinition.types[Path(path).suffix])

    def read(self) -> str:
        try:
            with open(self.path, mode='r', encoding='utf-8') as schema_file:
                file_content = schema_file.read()
            return file_content
        except (OSError, Exception):
            return ''

    @classmethod
    def parse(cls, scheme: "ConfigSchema") -> Dict[str, Cfgu]:
        """"""
        error_location = [cls.__name__, 'parse']
        # Read file as text
        schema_content = scheme.read()
        # parse per type
        if scheme.type == ConfigSchemaType.JSON:
            try:
                schema_content = json.loads(schema_content)
                schema_content = from_dict(Cfgu.from_dict, schema_content)
            except (JSONDecodeError, Exception):
                raise ValueError(error_message(f"Couldn't read or parse the file"))

        # validate parsed
        for key, cfgu in schema_content.items():
            if not is_valid_name(key):
                # todo `path nodes mustn't contain reserved words "{key}"` this is not a good suggestion
                raise ValueError(error_message(f"invalid key {key}", error_location + [key]))
            if cfgu.type == CfguType.REG_EX and cfgu.pattern is None:
                # todo suggestion grammar
                raise ValueError(error_message(f"invalid type property", error_location + [key, cfgu.type.value]),
                                 f'type {cfgu.type.value}" must come with a pattern property')
            if cfgu.default is not None:
                if cfgu.required is not None or cfgu.template is not None:
                    # todo suggestion grammar
                    raise ValueError(error_message(f"invalid default property", error_location + [key, 'default']),
                                     f"default mustn't set together with required or template properties")
                if cfgu.type == CfguType.REG_EX:
                    if re.fullmatch(cfgu.pattern, cfgu.default) is None:
                        raise ValueError(error_message(f"invalid default property", error_location + [key, 'default']),
                                         f"{cfgu.default} doesn't match {cfgu.pattern}")
                else:
                    type_test = ConfigSchema._SchemaDefinition.VALIDATORS.get(cfgu.type.value, lambda: False)
                    if not type_test(cfgu.default):
                        raise ValueError(error_message(f"invalid default property", error_location + [key, 'default']),
                                         f"{cfgu.default} must be of type {cfgu.type.value}")

            if cfgu.depends is not None and (not len(cfgu.depends)
                                             or any([not is_valid_name(dependency) for dependency in cfgu.depends])):
                raise ValueError(error_message(f"invalid depends property", error_location + [key, 'depends']),
                                 f"depends is empty or contain reserved words")

        return schema_content

File: go/core/config_schema.go
package core

import (
	"fmt"
	"os"
	osPath "path"
	"regexp"
	"strings"

	configu "github.com/configu/configu/go"

	"github.com/configu/configu/go/core/generated"

	"github.com/go-playground/validator/v10"
)

const (
	EXT string = ".cfgu"
)

type ConfigSchema generated.ConfigSchema

func NewConfigSchema(path string) (schema ConfigSchema, err error) {
	extensions := strings.Join([]string{
		fmt.Sprintf("%s.%s", EXT, generated.JSON),
	},
		" | ")
	pattern := fmt.Sprintf(".*(%s)", extensions)
	if match, _ := regexp.MatchString(pattern, path); !match {
		err = configu.ConfiguError{
			Message:    fmt.Sprintf("invalid path %s", path),
			Location:   []string{"core", "NewConfigSchema"},
			Suggestion: fmt.Sprintf("file extension must be %s", extensions),
		}
		return
	}
	schema = ConfigSchema{

		Path: path,
		Type: generated.ConfigSchemaType(osPath.Ext(path)),
	}
	return
}

func (schema ConfigSchema) Read() []byte {
	content, err := os.ReadFile(schema.Path)
	if err != nil {
		return []byte{}
	}
	return content
}

func (schema ConfigSchema) Parse() (cfgu generated.ConfigSchemaContents, err error) {
	content := schema.Read()
	schema_content, e := generated.UnmarshalConfigSchemaContents(content)
	if e != nil {
		err = configu.ConfiguError{
			Message: "Couldn't parse schema file",
		}
		return
	}
	for key, value := range schema_content {
		if !configu.IsValidName(key) {
			err = configu.ConfiguError{
				Message:    fmt.Sprintf("invalid key %s", key),
				Location:   []string{"core", "ConfigSchema", "Parse", key},
				Suggestion: fmt.Sprintf("path nodes mustn't contain reserved word %s", key),
			}
			return
		}
		if value.Type == generated.RegEx && value.Pattern == nil {
			err = configu.ConfiguError{
				Message:    "invalid type property",
				Location:   []string{"core", "ConfigSchema", "Parse", key, string(value.Type)},
				Suggestion: fmt.Sprintf("type %s must come with a pattern property", value.Type),
			}
			return
		}
		if value.Default != nil {
			if value.Required != nil || value.Template != nil {
				err = configu.ConfiguError{
					Message:    "invalid default property",
					Location:   []string{"core", "ConfigSchema", "Parse", key, "default"},
					Suggestion: "default mustn't set together with required or template properties",
				}
				return
			} else {
				if !ValidateCfguType(value, *value.Default) {
					err = configu.ConfiguError{
						Message:    "invalid default property",
						Location:   []string{"core", "ConfigSchema", "Parse", key, "default"},
						Suggestion: fmt.Sprintf("%v must be of type %v or match Regex", value.Default, value.Type),
					}
					return
				}
			}
		}
		if value.Depends != nil {
			if len(value.Depends) == 0 || anyInvalidDependencyNames(value) {
				err = configu.ConfiguError{
					Message:    "invalid depends property",
					Location:   []string{"core", "ConfigSchema", "Parse", key, "depends"},
					Suggestion: "depends is empty or contains reserved words",
				}
				return
			}
		}
	}
	cfgu = schema_content
	return
}

func anyInvalidDependencyNames(value generated.ConfigSchemaContentsValue) bool {
	for _, dependency := range value.Depends {
		if !configu.IsValidName(dependency) {
			return false
		}
	}
	return true
}

func ValidateCfguType(schema generated.ConfigSchemaContentsValue, value string) bool {
	switch schema.Type {
	case generated.Base64:
		return isValidBase64(value)
	case generated.Boolean:
		return isValidBoolean(value)
	case generated.Color:
		return isValidColor(value)
	case generated.ConnectionString:
		return isValidConnectionString(value)
	case generated.Country:
		return isValidCountry(value)
	case generated.Currency:
		return isValidCurrency(value)
	case generated.Domain:
		return isValidDomain(value)
	case generated.Email:
		return isValidEmail(value)
	case generated.Hex:
		return isValidHex(value)
	case generated.IPv4:
		return isValidIPv4(value)
	case generated.IPv6:
		return isValidIPv6(value)
	case generated.LatLong:
		return isValidLatLong(value)
	case generated.Md5:
		return isValidMd5(value)
	case generated.Number:
		return isValidNumber(value)
	case generated.RegEx:
		return isValidRegEx(value, *schema.Pattern)
	case generated.SHA:
		return isValidSHA(value)
	case generated.SemVer:
		return isValidSemVer(value)
	case generated.String:
		return isValidString(value)
	case generated.URL:
		return isValidURL(value)
	case generated.UUID:
		return isValidUUID(value)
	default:
		return false
	}
}

func isValidBase64(value string) bool {
	type MyStruct struct {
		Value string `validate:"required,base64"`
	}
	return validator.New().Struct(MyStruct{value}) == nil
}

func isValidBoolean(value string) bool {
	type MyStruct struct {
		Value string `validate:"required,boolean"`
	}
	return validator.New().Struct(MyStruct{value}) == nil
}

func isValidColor(value string) bool {
	type MyStruct struct {
		Value string `validate:"required,iscolor"`
	}
	return validator.New().Struct(MyStruct{value}) == nil
}

func isValidConnectionString(value string) bool {
	return regexp.MustCompile(`^(?:([^:/?#\s]+):/{2})?(?:([^@/?#\s]+)@)?([^/?#\s]+)?(?:/([^?#\s]*))?(?:[?]([^#\s]+))?\S*$`).MatchString(value)
}

func isValidCountry(value string) bool {
	type MyStruct struct {
		Value string `validate:"required,country_code"`
	}
	return validator.New().Struct(MyStruct{value}) == nil
}

func isValidCurrency(value string) bool {
	type MyStruct struct {
		Value string `validate:"required,iso4217"`
	}
	return validator.New().Struct(MyStruct{value}) == nil
}

func isValidDomain(value string) bool {
	type MyStruct struct {
		Value string `validate:"required,fqdn"`
	}
	return validator.New().Struct(MyStruct{value}) == nil
}

func isValidEmail(value string) bool {
	type MyStruct struct {
		Value string `validate:"required,email"`
	}
	return validator.New().Struct(MyStruct{value}) == nil
}

func isValidHex(value string) bool {
	type MyStruct struct {
		Value string `validate:"required,hexadecimal"`
	}
	return validator.New().Struct(MyStruct{value}) == nil
}

func isValidIPv4(value string) bool {
	type MyStruct struct {
		Value string `validate:"required,ipv4"`
	}
	return validator.New().Struct(MyStruct{value}) == nil
}

func isValidIPv6(value string) bool {
	type MyStruct struct {
		Value string `validate:"required,ipv6"`
	}
	return validator.New().Struct(MyStruct{value}) == nil
}

func isValidLatLong(value string) bool {
	if !strings.Contains(value, ",") {
		return false
	}
	parts := strings.Split(value, ",")
	if len(parts) != 2 {
		return false
	}
	if strings.HasPrefix(parts[0], "(") && !strings.HasSuffix(parts[1], ")") {
		return false
	}
	if !strings.HasPrefix(parts[0], "(") && strings.HasSuffix(parts[1], ")") {
		return false
	}
	type MyStruct struct {
		Latitude  string `validate:"required,latitude"`
		Longitude string `validate:"required,longitude"`
	}
	return validator.New().Struct(MyStruct{parts[0], parts[1]}) == nil
}

func isValidMd5(value string) bool {
	type MyStruct struct {
		Value string `validate:"required,md5"`
	}
	return validator.New().Struct(MyStruct{value}) == nil
}

func isValidNumber(value string) bool {
	type MyStruct struct {
		Value string `validate:"required,number"`
	}
	return validator.New().Struct(MyStruct{value}) == nil
}

func isValidRegEx(value string, pattern string) bool {
	re, err := regexp.Compile(pattern)
	if err != nil {
		return false
	}
	return re.FindString(value) != ""
}

func isValidSHA(value string) bool {
	type MyStruct struct {
		Value string `validate:"required,sha256"`
	}
	return validator.New().Struct(MyStruct{value}) == nil
}

func isValidSemVer(value string) bool {
	type MyStruct struct {
		Value string `validate:"required,semver"`
	}
	return validator.New().Struct(MyStruct{value}) == nil
}

func isValidString(value string) bool {
	type MyStruct struct {
		Value string `validate:"required"`
	}
	return validator.New().Struct(MyStruct{value}) == nil
}

func isValidURL(value string) bool {
	type MyStruct struct {
		Value string `validate:"required,datauri"`
	}
	return validator.New().Struct(MyStruct{value}) == nil
}

func isValidUUID(value string) bool {
	type MyStruct struct {
		Value string `validate:"required,uuid"`
	}
	return validator.New().Struct(MyStruct{value}) == nil
}
